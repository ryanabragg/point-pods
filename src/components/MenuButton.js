import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';

import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import MenuList from '@material-ui/core/MenuList';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grow from '@material-ui/core/Grow';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';

const styles = theme => ({
  root: {
    display: 'flex',
  },
  paper: {
    marginRight: theme.spacing.unit * 2,
  },
});

class MenuButton extends React.Component {
  state = {
    open: false,
  };

  menuAnchor = null;

  handleToggle = () => this.setState(state => ({ open: !state.open }));

  handleClose = event => {
    if(this.menuAnchor.contains(event.target))
      return;
    this.setState({ open: false });
  };

  handleClick = (i) => (event) => {
    if(typeof this.props.menuItems[i].action === 'function')
      this.props.menuItems[i].action();
  }

  render() {
    const { classes, buttonType, buttonProps, buttonContent, menuItems } = this.props;
    const { open } = this.state;
    let TheButton;
    switch(buttonType) {
      default:
      case 'text': TheButton = Button; break;
      case 'icon': TheButton = IconButton; break;
    }

    return (
      <div className={classes.root}>
          <TheButton
            {...buttonProps}
            buttonRef={node => this.menuAnchor = node}
            aria-owns={open ? 'button-menu' : null}
            aria-haspopup='true'
            onClick={this.handleToggle}
          >
            {buttonContent}
          </TheButton>
          <Popper open={open} anchorEl={this.menuAnchor} transition disablePortal>
            {({ TransitionProps, placement }) => (
              <Grow
                {...TransitionProps}
                id='button-menu'
                style={{ transformOrigin: placement === 'bottom' ? 'center top' : 'center bottom' }}
              >
                <Paper>
                  <ClickAwayListener onClickAway={this.handleClose}>
                    <MenuList onClick={this.handleClose}>
                      {menuItems.map((item, i) => (
                        <MenuItem key={i}
                          onClick={this.handleClick(i)}
                        >
                          {item.icon && (
                            <ListItemIcon>
                              {item.icon}
                            </ListItemIcon>
                          )}
                          {item.icon
                            ? <ListItemText inset primary={item.label} />
                            : item.label
                          }
                        </MenuItem>
                      ))}
                    </MenuList>
                  </ClickAwayListener>
                </Paper>
              </Grow>
            )}
          </Popper>
      </div>
    );
  }
}

MenuButton.defaultProps = {
  buttonType: 'normal',
  buttonContent: 'button',
  buttonProps: {},
  menuItems: [],
}

MenuButton.propTypes = {
  classes: PropTypes.object.isRequired, // added by withStyles
  buttonType: PropTypes.oneOf(['normal', 'icon']),
  buttonContent: PropTypes.node,
  buttonProps: PropTypes.object,
  menuItems: PropTypes.array,
};

export default withStyles(styles)(MenuButton);
