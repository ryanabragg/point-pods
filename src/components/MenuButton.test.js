import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Popper from '@material-ui/core/Popper';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import MenuList from '@material-ui/core/MenuList';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';

import MoreVertIcon from '@material-ui/icons/MoreVert';

import MenuButton from './MenuButton';

const items = [
  { label: 'One', action: jest.fn(v => 'One') },
  { label: 'Two', action: jest.fn(v => 'Two') },
  { label: 'Three', action: jest.fn(v => 'Three') },
];

const itemsWithIcons = [
  { label: 'One', icon: <MoreVertIcon />, action: jest.fn(v => 'One') },
  { label: 'Two', icon: <MoreVertIcon />, action: jest.fn(v => 'Two') },
  { label: 'Three', icon: <MoreVertIcon />, action: jest.fn(v => 'Three') },
];

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MenuButton />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches the prior snapshot', () => {
  const component = renderer.create(<MenuButton />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<MenuButton buttonContent='test' />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<MenuButton buttonProps={{color: 'inherit'}} buttonContent={<MoreVertIcon />} />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<MenuButton menuItems={items} />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<MenuButton menuItems={itemsWithIcons} />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('menu button', () => {
  const component = shallow(<MenuButton />).dive();
  expect(component.find(Button).length).toBe(1);
  expect(component.find(IconButton).length).toBe(0);
  expect(component.find(Button).prop('children')).toBe('button');
  component.setProps({
    buttonContent: 'test',
  });
  expect(component.find(Button).prop('children')).toBe('test');
  component.setProps({
    buttonType: 'icon',
    buttonContent: <MoreVertIcon />,
    buttonProps: { color: 'inherit' },
  });
  expect(component.find(Button).length).toBe(0);
  expect(component.find(IconButton).length).toBe(1);
  expect(component.find(IconButton).prop('children')).toEqual(<MoreVertIcon />);
  expect(component.find(IconButton).prop('color')).toBe('inherit');
  expect(component.state('open')).toBe(false);
  component.find(IconButton).simulate('click');
  expect(component.state('open')).toBe(true);
});

test('menu list'/*, () => {
  const component = shallow(<MenuButton menuItems={items} />).dive();
  component.find(Button).simulate('click');
  console.log(component.props())
  expect(component.find(MenuItem).length).toBe(items.length);
  component.setProps({
    menuItems: itemsWithIcons,
  });
  expect(component.find(ListItemIcon).length).toBe(itemsWithIcons.length);
  expect(component.find(ListItemText).length).toBe(itemsWithIcons.length);
}*/);

test('className prop', () => {
  const component = shallow(<MenuButton />).dive();
  expect(component.find('div').prop('className')).toBe('');
  component.setProps({className: 'test'})
  expect(component.find('div').prop('className')).toBe('test');
});

test('disablePortal prop pass-through', () => {
  const component = shallow(<MenuButton />).dive();
  expect(component.find(Popper).prop('disablePortal')).toBe(false);
  component.setProps({
    disablePortal: true,
  });
  expect(component.find(Popper).prop('disablePortal')).toBe(true);
});

test('paperStyle prop pass-through'/*, () => {
  const component = shallow(<MenuButton buttonContent='test' menuItems={items}/>).dive();
  component.setState({open: true});
  component.update();
  expect(component.find(Paper).prop('className')).toBe('');
  const style = 'maxHeight: 200; overflow: auto';
  component.setProps({
    paperStyle: style,
  });
  expect(component.find(Paper).prop('className')).toBe(style);
}*/);
