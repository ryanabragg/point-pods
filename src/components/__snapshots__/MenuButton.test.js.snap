// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the prior snapshot 1`] = `
<div
  className=""
>
  <button
    aria-haspopup="true"
    aria-owns={null}
    className="MuiButtonBase-root-61 MuiButton-root-37 MuiButton-text-39 MuiButton-flat-42"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-38"
    >
      button
    </span>
    <span
      className="MuiTouchRipple-root-64"
    />
  </button>
</div>
`;

exports[`matches the prior snapshot 2`] = `
<div
  className=""
>
  <button
    aria-haspopup="true"
    aria-owns={null}
    className="MuiButtonBase-root-61 MuiButton-root-37 MuiButton-text-39 MuiButton-flat-42"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-38"
    >
      test
    </span>
    <span
      className="MuiTouchRipple-root-64"
    />
  </button>
</div>
`;

exports[`matches the prior snapshot 3`] = `
<div
  className=""
>
  <button
    aria-haspopup="true"
    aria-owns={null}
    className="MuiButtonBase-root-61 MuiButton-root-37 MuiButton-text-39 MuiButton-flat-42 MuiButton-colorInherit-56"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-38"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-71"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root-64"
    />
  </button>
</div>
`;

exports[`matches the prior snapshot 4`] = `
<div
  className=""
>
  <button
    aria-haspopup="true"
    aria-owns={null}
    className="MuiButtonBase-root-61 MuiButton-root-37 MuiButton-text-39 MuiButton-flat-42"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-38"
    >
      button
    </span>
    <span
      className="MuiTouchRipple-root-64"
    />
  </button>
</div>
`;

exports[`matches the prior snapshot 5`] = `
<div
  className=""
>
  <button
    aria-haspopup="true"
    aria-owns={null}
    className="MuiButtonBase-root-61 MuiButton-root-37 MuiButton-text-39 MuiButton-flat-42"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-38"
    >
      button
    </span>
    <span
      className="MuiTouchRipple-root-64"
    />
  </button>
</div>
`;
